### PR TITLE
fix IP buffer padding check on 64bit

### PR DESCRIPTION
**Pull request by twpedersen**
On 64 bit systems, FreeRTOS_IPInit() would assert
ipconfigBUFFER_PADDING was equal to 14 to "make sure there
is enough space in pucEthernetBuffer to store a pointer."

This prevents the driver from requesting additional
padding, so make the assert greater than or equal to 14.

Also use the final ipBUFFER_PADDING value instead of
ipconfigBUFFER_PADDING, which is probably what was
intended?

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by AniruddhaKanhere** [comment]
@twpedersen, thank you for taking the time to point this out and suggest a possible solution.
While I agree with your statement that we should be using `ipBUFFER_PADDING` since that gets the value of `ipconfigBUFFER_PADDING` when it is non zero, I wonder whether using a `>=` is correct. Let me try to explain this statement :)

We want enough space for storing the pointers, true, but we do need a particular buffer offset to get a 32-bit alignment of the buffer. See the comment [here](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/main/include/FreeRTOS_IP.h#L89). Therefore, using anything `>=14` (let's say 15) would defeat the purpose of keeping the buffer alligned.
Additionally, there should enough space for us to store pointers and the padding but we do want to keep it at the lower end of the spectrum since FreeRTOS+TCP is designed to be lightweight and less RAM/storage intensive.

@htibosch, comments/corrections?

**Comment by htibosch** [comment]
@twpedersen : thanks for noting this, but I did not want to test `ipBUFFER_PADDING`.

`ipBUFFER_PADDING` is defined as:
~~~c
#if ( ipconfigBUFFER_PADDING != 0 )
    #define ipBUFFER_PADDING    ipconfigBUFFER_PADDING
#else
    #define ipBUFFER_PADDING    ( 8U + ipconfigPACKET_FILLER_SIZE )
#endif
~~~
The second definition of `ipBUFFER_PADDING` can not be used on a 64-bit platform:
`ipconfigPACKET_FILLER_SIZE` it the offset from word-aligned that is used by DMA. As far as I know, all projects have it defined as 2, making a total of 10 bytes. You can not change it to become 14 bytes long.

`ipconfigBUFFER_PADDING` must be bigger than or equal to 14:
~~~c
    configASSERT( ( ipconfigBUFFER_PADDING ) >= 14 );
	/* But it must have this strange alignment: */
    configASSERT( ( ( ipconfigBUFFER_PADDING ) + 2 ) == 0 );
~~~

There are some drivers who use this hidden space to store some driver information. The layout of this space is explained [here](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/main/include/FreeRTOS_IP.h#L85).

So I would continue to check `ipconfigBUFFER_PADDING`, and maybe add some more comment.
Also we could test for this special alignment.

File: FreeRTOS_IP.c [hunk 1]
```diff
@@ -1150,7 +1150,7 @@ BaseType_t FreeRTOS_IPInit( const uint8_t ucIPAddress[ ipIP_ADDRESS_LENGTH_BYTES
     {
         /* This is a 64-bit platform, make sure there is enough space in
          * pucEthernetBuffer to store a pointer. */
-        configASSERT( ipconfigBUFFER_PADDING == 14 );
+        configASSERT( ipBUFFER_PADDING >= 14 );
```
Comment (AniruddhaKanhere) [inline FreeRTOS_IP.c:1153]:
```suggestion
        configASSERT( ipconfigBUFFER_PADDING >= 14 );

        /* But it must have this strange alignment: */
        configASSERT( ( ( ( ipconfigBUFFER_PADDING ) + 2 ) % 4 ) == 0 );
```

**Review by AniruddhaKanhere — CHANGES_REQUESTED** [review]
@twpedersen, Would you mind making the changes below if you do not have any objections/suggestions?

**Comment by AniruddhaKanhere** [comment]
Hello @twpedersen,
I am adding the suggested changes.

File: FreeRTOS_IP.c [hunk 2]
```diff
@@ -1140,7 +1140,10 @@ BaseType_t FreeRTOS_IPInit( const uint8_t ucIPAddress[ ipIP_ADDRESS_LENGTH_BYTES
     {
         /* This is a 64-bit platform, make sure there is enough space in
          * pucEthernetBuffer to store a pointer. */
-        configASSERT( ipconfigBUFFER_PADDING == 14 );
+        configASSERT( ipconfigBUFFER_PADDING >= 14 );
+
+        /* But it must have this strange alignment: */
+        configASSERT( ( ( ( ipconfigBUFFER_PADDING ) + 2 ) % 4 ) == 0 );
```
Comment (htibosch) [inline FreeRTOS_IP.c:1146]: Thanks Aniruddha for adding these changes. I think it is all good now.

**Review by htibosch — APPROVED** [review]

**Review by AniruddhaKanhere — APPROVED** [review]

File: FreeRTOS_IP.c [hunk 3]
```diff
@@ -1140,7 +1140,10 @@ BaseType_t FreeRTOS_IPInit( const uint8_t ucIPAddress[ ipIP_ADDRESS_LENGTH_BYTES
     {
         /* This is a 64-bit platform, make sure there is enough space in
          * pucEthernetBuffer to store a pointer. */
-        configASSERT( ipconfigBUFFER_PADDING == 14 );
+        configASSERT( ipconfigBUFFER_PADDING >= 14 );
+
+        /* But it must have this strange alignment: */
```
Comment (aggarg) [inline FreeRTOS_IP.c:1145]: Possible to explain why it must have this strange alignment?

**Review by aggarg — APPROVED** [review]

**Pull request merged**